### PR TITLE
Account for nibabel 5.0.0 removal of .py3k shim - use numpy.compat.py3k

### PR DESCRIPTION
**Pull request by musicinmybrain**
Rebased from:

https://salsa.debian.org/med-team/nipy/-/blob/12a4fbea8c99c1e5dc07ee81bc3da1a450617050/debian/patches/nibabel5.0.0.patch

This is a version of https://github.com/nipy/nipy/pull/498 that applies cleanly and seems to work as expected in Fedora Linux. Please see the original PR for discussion.

### CODE REVIEW

File: nipy/algorithms/statistics/formula/tests/test_formula.py [hunk 1]
```diff
@@ -18,7 +18,7 @@
 from ..formulae import terms, Term
 from nipy.utils import VisibleDeprecationWarning
 
-from nibabel.py3k import asbytes
+from numpy.compat.py3k import asbytes
```
Comment (effigies) [inline nipy/algorithms/statistics/formula/tests/test_formula.py:21]: IIRC numpy is deprecating this as well. We should just change `asbytes(s)` to `s.encode()`.

File: nipy/io/tests/test_image_io.py [hunk 2]
```diff
@@ -45,7 +45,7 @@ def test_badfile():
     # nibabel prior 2.1.0 was throwing a ImageFileError for the not-recognized
     # file type.  >=2.1.0 give a FileNotFoundError.
     try:
-        from nibabel.py3k import FileNotFoundError
+        from numpy.compat.py3k import FileNotFoundError
     except ImportError:
         FileNotFoundError = IOError
```
Comment (effigies) [inline nipy/io/tests/test_image_io.py:50]: Should just remove this. `FileNotFoundError` is now a builtin.

File: tools/run_log_examples.py [hunk 3]
```diff
@@ -25,7 +25,7 @@
 from subprocess import Popen, PIPE
 import re
 
-from nibabel.py3k import asstr
+from numpy.compat.py3k import asstr
```
Comment (effigies) [inline tools/run_log_examples.py:28]: Likewise, here we should just convert `asstr(s)` to `s.decode()`

File: nipy/algorithms/statistics/formula/tests/test_formula.py (hunk 1 above)
Comment (musicinmybrain) [reply nipy/algorithms/statistics/formula/tests/test_formula.py:21]: Seems logical. Adjusted.

File: nipy/io/tests/test_image_io.py (hunk 2 above)
Comment (musicinmybrain) [reply nipy/io/tests/test_image_io.py:50]: I assume this package is no longer attempting to support Python 2.7?

Comment (effigies) [reply nipy/io/tests/test_image_io.py]: Given that there's hardly maintenance effort to keep it running, extending what little there is to 2.7 seems silly. I don't call the shots here, so I might get overruled, but it seems safe to drop Python 2.

**Review by effigies — APPROVED** [review]

File: nipy/io/tests/test_image_io.py (hunk 2 above)
Comment (matthew-brett) [reply nipy/io/tests/test_image_io.py]: Yup - no need to maintain Python 2.7 at this stage.

File: nipy/core/image/image_spaces.py [hunk 4]
```diff
@@ -60,7 +60,7 @@
 It also works with nibabel images, which can only have xyz_affines:
 
 >>> import nibabel as nib
->>> nimg = nib.Nifti1Image(data, affine)
+>>> nimg = nib.Nifti1Image(data.astype('float'), affine)
```
Comment (matthew-brett) [inline nipy/core/image/image_spaces.py:63]: Sorry - remind me why we need to cast as float now?

**Review by matthew-brett — COMMENTED** [review]
One point of clarification.

File: nipy/core/image/image_spaces.py (hunk 4 above)
Comment (effigies) [reply nipy/core/image/image_spaces.py:63]: It's a quick hack. The problem is that `np.arange()` (used to generate synthetic data) produces `int64`s and now we error on int64 if we don't get an explicit `header` or `dtype` parameter.

* https://github.com/nipy/nibabel/issues/1089

Comment (matthew-brett) [reply nipy/core/image/image_spaces.py:63]: Worth a comment to that effect?

Comment (effigies) [reply nipy/core/image/image_spaces.py:63]: I think a comment in a docstring would be more distracting than illuminating. Better solution, IMO, would be to update the `data = np.arange(...)` to be `data = np.arange(..., dtype='float32')`.

Comment (matthew-brett) [reply nipy/core/image/image_spaces.py:63]: Yes, that would also be fine.

Comment (musicinmybrain) [reply nipy/core/image/image_spaces.py]: I pushed a follow-up commit that I think should be satisfactory.

**Pull request merged**